crates/client/block_import/src/verify_apply.rs

```diff
@@ -293,3 +294,635 @@ fn block_hash(
+            let backend = setup_test_backend().await;
+            let header = create_dummy_header();
+            let pending_block = finalized_block_zero(header);
+            backend.store_block(pending_block.clone(), finalized_state_diff_zero(), vec![]).unwrap();
+
+            let parent_block_hash = Some(felt!("0x1")); // Mismatch
+            let unverified_block_number = pending_block.info.block_n().map(|n| n + 1);
+            let validation = create_validation_context(false);
+
+            let result = check_parent_hash_and_num(&backend, parent_block_hash, unverified_block_number, &validation);
+            assert!(matches!(result, Err(BlockImportError::ParentHash { .. })));
+        }
+
+        /// Test ignore_block_order flag functionality.
+        ///
+        /// Verifies that:
+        /// 1. When ignore_block_order is true, mismatches in block number and parent hash are allowed.
+        /// 2. The function returns the provided values instead of throwing errors.
+        #[tokio::test]
+        async fn test_check_parent_hash_and_num_ignore_block_order() {
+            let backend = setup_test_backend().await;
+            let header = create_dummy_header();
+            let pending_block = finalized_block_zero(header);
+            backend.store_block(pending_block.clone(), finalized_state_diff_zero(), vec![]).unwrap();
+
+            let parent_block_hash = Some(felt!("0x1")); // Mismatch
+            let unverified_block_number = pending_block.info.block_n().map(|n| n + 2); // Mismatch
+            let validation = create_validation_context(true);
+
+            let result = check_parent_hash_and_num(&backend, parent_block_hash, unverified_block_number, &validation)
+                .expect("Failed to check parent hash and number");
+            assert_eq!(result, (pending_block.info.block_n().unwrap() + 2, pending_block.info.block_hash().unwrap()));
+        }
+    }
+
+    mod calculate_state_root_tests {
+        use starknet_api::felt;
+
+        use super::*;
+
+        /// Test successful state root calculation.
+        ///
+        /// Verifies that:
+        /// 1. The function correctly calculates the state root from non-zero inputs.
+        /// 2. The result is verified and differs from both input roots.
+        #[test]
+        fn test_calculate_state_root_success() {
+            let contracts_trie_root = felt!("0x123456");
+            let classes_trie_root = felt!("0x789abc");
+
+            let result = calculate_state_root(contracts_trie_root, classes_trie_root);
+
+            assert_eq!(
+                result,
+                felt!("0x6beb971880d4b4996b10fe613b8d49fa3dda8f8b63156c919077e08c534d06e"),
+                "State root should match"
+            );
+            assert_ne!(result, contracts_trie_root, "State root should not be equal to contracts_trie_root");
+            assert_ne!(result, classes_trie_root, "State root should not be equal to classes_trie_root");
```

Comment:
nitpick: is this necessary? considering we're hardcoding `0x6beb971880d4b4996b10fe613b8d49fa3dda8f8b63156c919077e08c534d06e`, there's only one way that's possible.

Reply:
right, removed